src/devices/Tca9548a/Tca9548A.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Device.I2c;
+using System.Device.Model;
+using System.IO;
+using Iot.Device.Board;
+
+namespace Iot.Device.Tca9548a
+{
+    /// <summary>
+    /// Tca9548A - 8-Channel I2C Switch/Multiplexer
+    /// </summary>
+    [Interface("BNO055 - 8-Channel I2C Switch/Multiplexer")]
+    public class Tca9548A : IDisposable
+    {
+        private readonly bool _shouldDispose;
+        private readonly I2cBus _i2cBus;
+
+        /// <summary>
+        /// The default I2C Address, page 15 of the main documentation
+        /// https://www.ti.com/lit/ds/symlink/tca9548a.pdf
+        /// </summary>
+        public const byte DefaultI2cAddress = 0x70;
+
+        /// <summary>
+        /// Array of all possible Multiplexer Channels
+        /// </summary>
+        public static readonly Channels[] DeviceChannels = (Channels[])Enum.GetValues(typeof(Channels));
+
+        private I2cDevice _i2CDevice;
+
+        /// <summary>
+        /// Creates a Multiplexer Instance
+        /// </summary>
+        /// <param name="i2CBus">The I2C Bus on which Mux is</param>
+        /// <param name="address"> Address of the Mux on I2C bus</param>
+        /// <param name="shouldDispose">true to dispose the I2C device at dispose</param>
+        /// <exception cref="ArgumentNullException">Exception thrown if I2C device is null</exception>
+        public Tca9548A(I2cBus i2CBus, byte address = DefaultI2cAddress, bool shouldDispose = true)
```

Comment:
why not passing directly the I2cDevice instead of the I2cBus?

Reply:
Another design alternative:

```csharp
public Tca9548A : IList<I2cBus>
{
  public Tca9548A(I2cBus parentBus) {...}
  public T this[int i] => new Tc29548Bus(_parentBus, channel); // even better store it before hand in the private list and do not create it every time, this is just a concept idea
  // ...
}
```

```csharp
public Tca9548ABus : I2cBus // entire class can be internal if it doesn't have anything special over I2cBus
{
  internal Tca9548ABus(Tca9548A parent, int channel) {...}
  public I2cDevice CreateDevice(int address) => new Tca9548AChannel(...);
}
```

```csharp
public class Tca9548AChannel : I2cDevice
{
    internal Tca9548AChannel(Tca9548ABus parent, int channel) { ...}
    
    public override void Write(ReadOnlySpan<byte> buffer)
    {
        SelectChannel(...);
        // do the writing
    }
   // etc...
}   
```

Reply:
@krwq  That's actually what I suggested already in my post. That is my preferable solution.

Reply:
@Ellerbach Because it needs both: The bus (from the main controller) and the device (of the TCA itself). That's because one can create client devices from the bus, but not determine the bus from a device. The suggested implementation in `I2cBusExtensions` (see individual commits) does not work.

Reply:
Hi @krwq, I am still not able to understand how the Tca9548AChannel object will select channel on the TCA9548A device in this implementation which is on different i2c address when we have to read or write. 
```

public class Tca9548AChannel : I2cDevice
{
    internal Tca9548AChannel(Tca9548ABus parent, int channel) { ...}
    
    public override void Write(ReadOnlySpan<byte> buffer)
    {
        SelectChannel(...);
        // do the writing
    }   
}   
```

Reply:
Maybe the names are not optimal yet (maybe `Tca9548AChannel` should better be named `Tca9548ADevice`). 

The idea is that if you do `channel1Bus.CreateDevice(Bmp180.DefaultI2cAddress)` (with `channel1Bus` being an instance of a `Tca9548ABus`), then you get an instance of a `Tca9548AChannel`, which is a subclass of `I2cDevice`.  When the client binding now uses this instance (of which he doesn't know the exact type) the channel selection happens behind the scene, because of this implementation:

```csharp
public override void Write(ReadOnlySpan<byte> buffer)
    {
        SelectChannel(_channel);
        _realBus.Write(buffer);
    }   
``` 

This basically first selects the channel (which is a member of this class) and then forwards the call to the underlying I2C controller for the actual write. So this secretly inserts a channel selection command before every write command. (Of course, it would be useful to cache the currently selected channel as to avoid setting the same value over and over again, but that's a detail).

Reply:
> But to select a channel via any device created by TCA bus, it needs a reference to the original TCA device at address say 0x70 , so do I reference the main TCA class in all the devices created

Reply:
@asheesh1996 the implementation detail doesn't matter. The idea is that user only sees I2cDevice and if it writes to that all magic happens behind the scenes